Support mapping basic type constants to blackboard inputs

bjoernschroeder opened the pull request.

bjoernschroeder committed:
set constant value from pml file

bjoernschroeder committed:
add check if value field exists for keymapping, fix order of variables for input mapping, update scheduling test to use constant bb values

bjoernschroeder committed:
include tuple, check that parentKey/childKey combination is not already stored in the inputMapping

veerajsk commented on alica_engine/include/engine/blackboard/KeyMapping.h:
```diff
@@ -11,12 +12,13 @@ class Blackboard;
 public:
-    using KeyMappingList = std::vector<std::pair<std::string, std::string>>;
+    // contains tuple of parentKey, childKey and value
+    using KeyMappingList = std::vector<std::tuple<std::string, std::string, std::string>>;
```
instead of a tuple would prefer to have a struct which conveys the intention clearly i.e.
```
struct Mapping {
    std::variant<std::string, std::string> parentKeyOrConstant;
    std::string childKey;
};
```

bjoernschroeder replied:
Are you sure with std::variant here? It seems like this should be std::pair instead, the first value of the pair is the parentKey, the second value is the constant.

veerajsk replied:
child key can be mapped to parent key or constant, but not both, therefore I went for std::variant

veerajsk commented on alica_engine/include/engine/blackboard/KeyMapping.h:
```diff
@@ -11,12 +12,13 @@ class Blackboard;
     const KeyMappingList& getInputMapping() const;
     const KeyMappingList& getOutputMapping() const;
-    void addInputMapping(const std::string& parentKey, const std::string& childKey);
+    void addInputMapping(const std::string& parentKey, const std::string& childKey, const std::string& value = "");
```
again, api should reflect the intention i.e.
`addInputMapping(const std::variant<std::string, std::string>& parentKeyOrConstant, const std::string& childKey)`

veerajsk commented on alica_engine/src/engine/blackboard/BlackboardUtil.cpp:
```diff
@@ -12,13 +12,24 @@ void BlackboardUtil::setInput(const Blackboard* parent_bb, Blackboard* child_bb,
+                childBb.setConstValue(childKey, value);
+                Logging::logDebug("BlackboardUtil") << "passing const value " << value << " into " << childKey;
+            } catch (std::exception& e) {
+                Logging::logError("BlackboardUtil") << "Blackboard error passing const value " << value << " into " << childKey << ". " << e.what();
```
rethrow the exception

veerajsk commented on alica_engine/src/engine/modelmanagement/factories/KeyMappingFactory.cpp:
```diff
@@ -12,8 +12,14 @@ std::unique_ptr<KeyMapping> KeyMappingFactory::create(const YAML::Node& node)
         for (const auto& entry : inputList) {
-            keyMapping->addInputMapping(
-                    Factory::getValue<std::string>(entry, alica::Strings::parentKey), Factory::getValue<std::string>(entry, alica::Strings::childKey));
+            // TODO: Remove ifelse once backend has been updated and all keymapping nodes have an (empty) value field
```
I don't follow this comment. The value field need not always be empty right? Either parent key or value needs to be filled

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/etc/plans/SchedulingTestPlan1.pml:
```diff
@@ -170,7 +171,8 @@
             "childKey": "numberOfCalls",
-            "parentKey": "Init2Term"
+            "parentKey": "",
+            "value": "5" # MANUALLY ADDED, NEEDS CHANGES TO BACKEND
```
I'd prefer to add a new test following the new test design

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -170,6 +170,20 @@ class BlackboardImpl
+        auto typeIndex = getTypeIndex(yamlTypeIt->second);
+        if (typeIndex.has_value()) {
+            _vals.at(targetKey) = makeBBValueForIndex<true>::make(typeIndex.value(), value);
+        }
```
else throw exception. This means that yaml type is not known or is std::any. In either case we don't know how to parse it

bjoernschroeder committed:
Merge branch 'devel' into bas_bb_map_constants

bjoernschroeder committed:
- add mapping struct
- rethrow exceptions
- throw exception if const value type is unknown or std::any

bjoernschroeder committed:
fix assert for output mapping

bjoernschroeder committed:
remove empty line, remove tuple include

bjoernschroeder committed:
use index to access variant value

bjoernschroeder committed:
check if defaultValue node exists

bjoernschroeder committed:
add tests for constant value, regenerate

bjoernschroeder committed:
remove comment

bjoernschroeder committed:
fix naming of plan elements

bjoernschroeder committed:
Merge branch 'devel' into bas_bb_map_constants

Dale-Koenig commented on alica_engine/include/engine/blackboard/KeyMapping.h:
```diff
@@ -1,22 +1,29 @@
 
+struct Mapping
+{
+    std::variant<std::string, std::string> parentKeyOrConstant;
```
imo we should define real types for each of these (i.e. a simple struct constructable for string) so that we can use the type to distinguish the argument

veerajsk replied:
okay, makes sense

veerajsk commented on alica_engine/src/engine/blackboard/KeyMapping.cpp:
```diff
@@ -19,16 +19,26 @@ const KeyMapping::KeyMappingList& KeyMapping::getOutputMapping() const
 {
-    assert(std::count(_inputMapping.begin(), _inputMapping.end(), std::make_pair(parentKey, childKey)) == 0);
-    _inputMapping.push_back(std::make_pair(parentKey, childKey));
+    assert(std::count_if(_inputMapping.begin(), _inputMapping.end(), [parentKeyOrConstant, childKey](Mapping& entry) {
```
I think you check if the childKey is already mapped & if so just return (with a error log printed) instead of assert

veerajsk commented on alica_engine/src/engine/blackboard/KeyMapping.cpp:
```diff
@@ -19,16 +19,26 @@ const KeyMapping::KeyMappingList& KeyMapping::getOutputMapping() const
+    std::variant<std::string, std::string> parentKeyVar;
+    parentKeyVar.emplace<0>(parentKey);
+
+    assert(std::count_if(_outputMapping.begin(), _outputMapping.end(), [parentKeyVar, childKey](Mapping& entry) {
```
analogous to previous comment

veerajsk commented on alica_engine/src/engine/modelmanagement/factories/BlackboardBlueprintFactory.cpp:
```diff
@@ -13,7 +13,7 @@ std::unique_ptr<BlackboardBlueprint> BlackboardBlueprintFactory::create(const YA
         auto type = getValue<std::string>(entry, Strings::stateType);
         std::optional<std::string> defaultValue;
-        if (!entry[Strings::defaultValue].IsNull()) {
+        if (entry[Strings::defaultValue] && !entry[Strings::defaultValue].IsNull()) {
```
there is no default value anymore

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/Expr/src/Behaviour/IncreaseCountByXBeh.cpp:
```diff
@@ -0,0 +1,47 @@
+{
+    /*PROTECTED REGION ID(initialiseParameters1084111613399827667) ENABLED START*/
+    // Add additional options here
+    CounterClass::called += LockedBlackboardRO(*getBlackboard()).get<uint64_t>("increaseBy");
```
we should not be using CounterClass anymore or any other statics. Best to use the global blackboard directly i.e. I don't think we even need to have a world model on the global blackboard, instead just use the global blackboard as the world model

veerajsk commented on supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/BlackboardTestPlan4610306063152670816.h:
```diff
@@ -0,0 +1,53 @@
+    /*PROTECTED REGION END*/
+};
+
+class UtilityFunction4610306063152670816 : public BasicUtilityFunction
```
need to update the PR to catch up to devel

Dale-Koenig replied:
Should be done after we finish getting rid of codegen relics

veerajsk replied:
I am going to open a new PR & close this one after we get rid of code gen in alica tests & use the new etc folder structure & folder location. Will be easier than fixing up this PR
